presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -838,54 +814,78 @@ private void queryFinished(ManagedQueryExecution query)
                 runningQueries.remove(query);
                 InternalResourceGroup group = this;
                 while (group.parent.isPresent()) {
-                    group.parent.get().descendantRunningQueries--;
+                    group.parent.get().descendantRunningQueries.decrementAndGet();
                     group = group.parent.get();
                 }
             }
             else {
                 queuedQueries.remove(query);
                 InternalResourceGroup group = this;
                 while (group.parent.isPresent()) {
-                    group.parent.get().descendantQueuedQueries--;
+                    group.parent.get().descendantQueuedQueries.decrementAndGet();
                     group = group.parent.get();
                 }
             }
             updateEligibility();
         }
     }
 
-    // Memory usage stats are expensive to maintain, so this method must be called periodically to update them
-    protected void internalRefreshStats()
+    private void updateWaitingQueuedQueries()
     {
-        checkState(Thread.holdsLock(root), "Must hold lock to refresh stats");
-        synchronized (root) {
-            if (subGroups.isEmpty()) {
-                cachedMemoryUsageBytes = 0;
-                for (ManagedQueryExecution query : runningQueries) {
-                    cachedMemoryUsageBytes += query.getUserMemoryReservation().toBytes();
+        // For leaf group, when no queries can run, all queued queries are waiting for resources on this resource group.
+        if (subGroups.isEmpty()) {
+            waitingQueuedQueries.set(queuedQueries.size());
+        }
+        else {
+            // For internal groups, when no queries can run, only queries that could run on its subgroups are waiting for resources on this group.
+            int waitingQueuedQueryCount = 0;
+            for (InternalResourceGroup subGroup : subGroups.values()) {
+                if (subGroup.canRunMore()) {
```

Comment:
Should we split out canRunMore also into a canRunMoreLocked method that assumes that the lock is held ?

Reply:
No, reading value should not need to call canRunMore(). 

The goal is to make sure any fb303 wants to read the stats do not need to block. 
And the stats should be able to be updated periodically in between the query scheduling cycle.